### PR TITLE
[Function] Refactor code of datetime comparetion and delete some unless code

### PR DESCRIPTION
**Pull request by HappenLee**
## Proposed changes

1. Refactored the logic of DateTime comparison size, and does not depend on the endianness of the type. Need to be compatible with the logic of the old version of doris, but it will cause the performance of DateTime comparison to decrease
2. Delete the useless code logic in the comparison function, and modify the format
3. Fixed the core problem of in/cast in function context

## Types of changes

What types of changes does your code introduce to Doris?
_Put an `x` in the boxes that apply_

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices apply)
- [x] Code refactor (Modify the code structure, format the code, etc...)
- [ ] Optimization. Including functional usability improvements and performance improvements.
- [ ] Dependency. Such as changes related to third-party components.
- [ ] Other.

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [ ] I have created an issue on (Fix #ISSUE) and described the bug/feature there in detail
- [ ] Compiling and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If these changes need document changes, I have updated the document
- [ ] Any dependent changes have been merged

## Further comments

If this is a relatively large or complex change, kick off the discussion at dev@doris.apache.org by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: be/src/vec/functions/functions_comparison.h [hunk 1]
```diff
@@ -768,46 +470,31 @@ class FunctionComparison : public IFunction {
...
         return std::make_shared<DataTypeUInt8>();
     }
 
     Status execute_impl(FunctionContext* context, Block& block, const ColumnNumbers& arguments,
                         size_t result, size_t input_rows_count) override {
-        const auto& col_with_type_and_name_left = block.get_by_position(arguments[0]);
-        const auto& col_with_type_and_name_right = block.get_by_position(arguments[1]);
-        const IColumn* col_left_untyped = col_with_type_and_name_left.column.get();
-        const IColumn* col_right_untyped = col_with_type_and_name_right.column.get();
+        const auto &col_with_type_and_name_left = block.get_by_position(arguments[0]);
```
Comment (wangbo) [inline be/src/vec/functions/functions_comparison.h:491]: 引用符号放到变量前面是标准写法吗，我之前一直以为是放到类型后面

File: be/src/vec/functions/functions_comparison.h [hunk 2]
```diff
@@ -99,313 +99,60 @@ struct NumComparisonImpl {
...
-                                                               b_n);
+        size_t size = a.size();
+        const auto* a_pos = a.data();
+        const auto* b_pos = b.data();
+        UInt8* c_pos = c.data();
+        const auto* a_end = a_pos + size;
 
-            prev_a_offset = a_offsets[i];
+        while (a_pos < a_end) {
+            *c_pos = Op::apply(*a_pos, *b_pos);
```
Comment (wangbo) [inline be/src/vec/functions/functions_comparison.h:118]: 我有两个问题，第一个是，这个循环能向量化吗，第二个是，对比默认的方式，性能提升大概多少

**Comment by wangbo** [comment]
整体datetime改动这个块代码上没问题了；functionContext部分辛苦硕哥@wangshuo128 有空看下吧

File: be/src/vec/functions/functions_comparison.h (hunk 1 above)
Comment (HappenLee) [reply be/src/vec/functions/functions_comparison.h:491]: 这个是我的格式问题，我改一下

File: be/src/vec/functions/functions_comparison.h (hunk 2 above)
Comment (HappenLee) [reply be/src/vec/functions/functions_comparison.h:118]: 1. 这个循环在 = ，！=的时候计算可以向量化，但是在 < > >= <=的时候不行
2. 对比默认的方式，datetime比较的性能是下降的，但是没办法，这个改动会导致兼容性问题，所以只能改回来

**Pull request merged**